src/hockeypuck/pghkp/storage.go

```diff
@@ -463,50 +661,431 @@ func (st *storage) insertKeyTx(tx *sql.Tx, key *openpgp.PrimaryKey) (isDuplicate
+	// (3) Insert single copy of in-file Duplicates, if they have no Duplicate in final keys table (in DB)
+	txStrs := []string{bulkTxFilterUniqueKeys, bulkTxPrepKeyStats, bulkTxFilterDupKeys}
+	msgStrs := []string{"bulkTx-filter-unique-keys", "bulkTx-prep-key-stats", "bulkTx-filter-dup-keys"}
+	err = st.bulkInsertSingleTx(txStrs, msgStrs)
+	if err != nil {
+		result.Errors = append(result.Errors, err)
+		return 0, false
+	}
+	return numNulls, true
+}
+
+func (st *storage) bulkInsertCheckedKeysSubkeys(keys []*openpgp.PrimaryKey,
+	result *hkpstorage.InsertError) (nullKeys, nullSubkeys int, ok bool) {
+	keysOK, subkeysOK := true, true
+	// key batch-processing
+	if nullKeys, keysOK = st.bulkInsertCheckKeys(result); !keysOK {
+		return 0, 0, false
+	}
+	// subkey batch-processing
+	if nullSubkeys, subkeysOK = st.bulkInsertCheckSubkeys(result); !subkeysOK {
+		return 0, 0, false
+	}
+
+	// Batch INSERT all checked-for-constraints keys from memory tables (should need no checks!!!!)
+	// Final batch-insertion in keys/subkeys tables without any checks: _must not_ give any errors
+	txStrs := []string{bulkTxInsertKeys, bulkTxInsertSubkeys}
+	msgStrs := []string{"bulkTx-insert-keys", "bulkTx-insert-subkeys"}
+	err := st.bulkInsertSingleTx(txStrs, msgStrs)
+	if err != nil {
+		result.Errors = append(result.Errors, err)
+		return 0, 0, false
+	}
+	return nullKeys, nullSubkeys, true
+}
+
+func (st *storage) bulkInsertSendBunchTx(keystmt, msgSpec string, keysValueArgs []interface{}) (err error) {
+	// In single transaction...
+	tx, err := st.Begin()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		} else {
+			tx.Commit()
 		}
+	}()
+
+	stmt, err := tx.Prepare(keystmt)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	_, err = stmt.Exec(keysValueArgs...) // All keys in bunch
+	if err != nil {
+		return errors.Wrapf(err, "cannot simply send a bunch of %s to server (too large bunch?)", msgSpec)
+	}
+	err = stmt.Close()
+	if err != nil {
+		return errors.Wrapf(err, "failed to close xfer sending a bunch of %s to server", msgSpec)
+	}
+
+	return nil
+}
 
-		if isDuplicate, err := st.insertKey(key); err != nil {
+func (st *storage) bulkInsertSendBunch(keystmt, subkeystmt string, keysValueArgs, subkeysValueArgs []interface{}) (err error) {
+
+	// Send all keys to in-mem tables to the pg server; *no constraints checked*
+	err = st.bulkInsertSendBunchTx(keystmt, "keys", keysValueArgs)
+	if err != nil {
+		return err
+	}
+
+	// Send all subkeys to in-mem tables to the pg server; *no constraints checked*
+	err = st.bulkInsertSendBunchTx(subkeystmt, "subkeys", subkeysValueArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type KeyInsertArgs struct {
+	RFingerprint *string
+	jsonStrDoc   *string
+	MD5          *string
+	keywords     *string
+}
+type SubkeyInsertArgs struct {
+	keyRFingerprint    *string
+	subkeyRFingerprint *string
+}
```

Comment:
Looks like these types are only used in private methods, so I think they could be private as well, keeping the public API smaller and more concise.

Reply:
OK. I will make them keyInsertArgs and subkeyInsertArgs.